packages/expo-image/src/ExpoImage.web.tsx

```diff
@@ -57,50 +62,149 @@ function getObjectPositionFromContentPosition(contentPosition?: ImageContentPosi
     .join(' ');
 }
 
-const ensureIsArray = <T extends any>(source: T | T[] | undefined) => {
+function ensureIsArray(source?: ImageSource): (ImageUriSource | RequireSource)[] {
   if (Array.isArray(source)) {
     return source;
   }
   if (source == null) {
     return [];
   }
   return [source];
+}
+
+type ImageState = 'empty' | 'loading' | 'loaded' | 'error';
+function useImageState(source: ImageSource | undefined) {
+  const [imageState, setImageState] = React.useState<ImageState>(source ? 'loading' : 'empty');
+  React.useEffect(() => {
+    setImageState((prevState) =>
+      prevState === 'empty' ? (source ? 'loading' : 'empty') : prevState
+    );
+  }, [source]);
+
+  const onLoad = React.useCallback(
+    () => setImageState((prevState) => (imageState === 'loading' ? 'loaded' : prevState)),
+    []
+  );
+  const handlers = React.useMemo(
+    () => ({
+      onLoad,
+    }),
+    [onLoad]
+  );
+  return [imageState, handlers] as [ImageState, { onLoad: () => void }];
+}
+
+const getCSSTiming = (timing?: ImageTransitionTiming) => {
+  return {
+    [ImageTransitionTiming.EASE_IN]: 'ease-in',
+    [ImageTransitionTiming.EASE_OUT]: 'ease-out',
+    [ImageTransitionTiming.EASE_IN_OUT]: 'ease-in-out',
+    [ImageTransitionTiming.LINEAR]: 'linear',
+  }[timing || ImageTransitionTiming.LINEAR];
```

Comment:
I think your solution is fine but only because of types and TS 😅 In plain JS this would work better if someone passes an invalid timing
```suggestion
  }[timing] ?? 'linear';
```